Don't fix access of an 'open' override in a 'public' extension

jrose-apple opened the pull request:
This is reasonable to diagnose with a warning, but dropping the 'open' down to 'public' isn't the right fix, because now it's not a valid override. The declaration has to get moved to another extension instead, or the extension has to not set a default access level.

This turned out to be a source compat issue because the same logic that emits the fix-it also updates the access of the member, which then resulted in "must be as accessible as the declaration it overrides" in the *same* build. It's not immediately clear what caused this; probably something's just being validated in a different order than it was before. The change makes sense either way.

rdar://problem/47557376
rdar://problem/28493971

jrose-apple committed:
Don't fix access of an 'open' decl in a 'public' extension

This is reasonable to diagnose with a warning, but dropping the 'open'
down to 'public' isn't the right fix, because now it's not a valid
override. The declaration has to get moved to another extension instead,
or the extension has to not set a default access level.

This turned out to be a source compat issue because the same logic
that emits the fix-it also updates the access of the member, which
then resulted in "must be as accessible as the declaration it
overrides" in the /same/ build. It's not immediately clear what caused
this; probably something's just being validated in a different order
than it was before. The change makes sense either way.

Stepping back, it's weird that a warning would change how the compiler
saw the code, and while we could check for 'override' explicitly, we
can't know if the member might be satisfying a protocol requirement.
Better to just not guess at the right answer here.

rdar://problem/47557376&28493971

jrose-apple commented:
@dingobye, you'd also be a good person to review, since you made changes here recently!

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

brentdax commented on test/ClangImporter/objc_override.swift:
```diff
@@ -116,6 +116,16 @@ class MyHashableNSObject: NSObject {
+  open override func method() {}
+}
+public extension OverrideInExtensionSub {
+  open override func accessWarning() {} // expected-warning {{declaring open instance method in a public extension}}
```
I know this diagnostic already exists, but "declaring open instance method in a public extension" isn't great. It doesn't really give you any idea of what the compiler will do with this code or how you should correct the issue. Perhaps the fix-it normally provides enough additional information to understand it, but there's no fix-it here. I think it should say something like "public keyword on extension will be ignored for this open instance method".

jrose-apple committed:
Tweak diagnostic for a high-access member in a low-access extension

Before: declaring a public instance method in a private extension
After: 'public' modifier conflicts with extension's default access of
       'private'

jrose-apple commented on test/ClangImporter/objc_override.swift:
```diff
@@ -116,6 +116,16 @@ class MyHashableNSObject: NSObject {
+  open override func method() {}
+}
+public extension OverrideInExtensionSub {
+  open override func accessWarning() {} // expected-warning {{'open' modifier conflicts with extension's default access of 'public'}}
```
@brentdax How does this sound? It's still vague about how it's going to be treated for now, but at least it sounds like a problem.

brentdax replied:
Much better!

jrose-apple commented:
[waiting for the previous tests to finish before starting more tests, since the only thing that changed was the diagnostic message]

jrose-apple commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -1640,10 +1640,15 @@ void AttributeChecker::visitAccessControlAttr(AccessControlAttr *attr) {
+        // (which can happen with Objective-C extensions). That can lead to a
+        // diagnostic fight between this and "declaration must be as accessible
+        // as the declaration it overrides".
+        if (!D->getAttrs().hasAttribute<OverrideAttr>()) {
```
Hrm. Maybe it'd be better to just remove the fix-it altogether. It's hard to tell whether something's trying to satisfy a requirement.

dingobye replied:
Agree! It is always better to be quiet than misleading.

jrose-apple replied:
We *will* still have the fighting cases for now in the error cases, but, well…that's not a regression.

dingobye replied:
Well, that is an orthogonal issue. 'type_witness_not_accessible_proto' keeps warning even though foo() is already declared `public`.

jrose-apple committed:
Remove mistaken early exit from access checking

Warnings should not result in early exits because we might end up
missing an error!

dingobye reviewed (approved):
LGTM!
